zerver/forms.py

```diff
@@ -253,51 +253,45 @@ class CreateUserForm(forms.Form):
     email = forms.EmailField()
 
 class OurAuthenticationForm(AuthenticationForm):
-    def clean_username(self):
-        # type: () -> str
-        email = self.cleaned_data['username']
-        try:
-            user_profile = get_user_profile_by_email(email)
-        except UserProfile.DoesNotExist:
-            return email
-
-        if user_profile.realm.deactivated:
-            error_msg = u"""Sorry for the trouble, but %s has been deactivated.
-
-Please contact %s to reactivate this group.""" % (
-                user_profile.realm.name,
-                FromAddress.SUPPORT)
-            raise ValidationError(mark_safe(error_msg))
-
-        if not user_profile.is_active and not user_profile.is_mirror_dummy:
-            error_msg = (
-                u"Your account is no longer active. "
-                u"Please contact your organization administrator to reactivate it.")
-            raise ValidationError(mark_safe(error_msg))
-
-        if not user_matches_subdomain(get_subdomain(self.request), user_profile):
-            logging.warning("User %s attempted to password login to wrong subdomain %s" %
-                            (user_profile.email, get_subdomain(self.request)))
-            raise ValidationError(mark_safe(WRONG_SUBDOMAIN_ERROR))
-        return email
-
     def clean(self):
         # type: () -> Dict[str, Any]
         username = self.cleaned_data.get('username')
         password = self.cleaned_data.get('password')
 
         if username is not None and password:
             subdomain = get_subdomain(self.request)
+            realm = get_realm(subdomain)
+            return_data = {}  # type: Dict[str, Any]
             self.user_cache = authenticate(self.request, username=username, password=password,
-                                           realm_subdomain=subdomain)
+                                           realm=realm, return_data=return_data)
+
+            if return_data.get("inactive_realm"):
+                raise AssertionError("Programming error: inactive realm in authentication form")
+
+            if return_data.get("inactive_user") and not return_data.get("is_mirror_dummy"):
+                # We exclude mirror dummy accounts here, because they
+                # should be treated as the user never having had an
+                # account (and thus will fall through to the normal
+                # invalid_login cas below.
```

Comment:
I don't totally follow this comment. (I see the actual logic hasn't changed, though, so this is just taking the opportunity to explain the slightly convoluted logic we already have. I appreciate that!)

Happily I did find "mirror dummy" [mentioned in our docs](http://zulip.readthedocs.io/en/latest/subsystems/users.html), with the docs' search bar!

Having read that, I think this comment would be fine with the logic (as well as spelling and punctuation) just a little cleaned up:

> We exclude mirror dummy accounts here. They should be treated as the user never having had an account, so we let them fall through to the normal `invalid_login` case below.

Reply:
Changed.